ms2/src/org/labkey/ms2/proteinCoverageMap.jsp

```diff
@@ -15,34 +15,163 @@
+        <option value="intensity" <%=isIntensityView ? h("selected") : h("")%> >Intensity</option>
+        <option value="confidenceScore" <%=isConfidenceView ? h("selected") : h("")%> >Confidence Score</option>
+    </select>
+</div>
+<%
 
+    List<Double> iValues = new ArrayList<>();
+    var peptideCharacteristics = bean.protein.getPeptideCharacteristics();
+    Map<Double, Color> heatMapColorRGB = new HashMap<>();
+
+    if (isIntensityView)
+    {
+        peptideCharacteristics.sort((o1, o2) -> {
+            if (o1.getIntensity() == null && o2.getIntensity() == null) return 0;
+            if (o2.getIntensity() == null) return 1;
+            if (o1.getIntensity() == null) return -1;
+            return o2.getIntensity().compareTo(o1.getIntensity());
+        });
+        peptideCharacteristics.forEach(peptideCharacteristic -> {
+            if (peptideCharacteristic.getIntensity() != null)
+            {
+                iValues.add(peptideCharacteristic.getIntensity());
+            }
+        });
+    }
+    if (isConfidenceView)
+    {
+        peptideCharacteristics.sort((o1, o2) -> {
+            if (o1.getConfidence() == null && o2.getConfidence() == null) return 0;
+            if (o2.getConfidence() == null) return 1;
+            if (o1.getConfidence() == null) return -1;
+            return o2.getConfidence().compareTo(o1.getConfidence());
+        });
+        peptideCharacteristics.forEach(peptideCharacteristic -> {
+          if (peptideCharacteristic.getConfidence() != null)
+          {
+              iValues.add(peptideCharacteristic.getConfidence());
+          }
+        });
+    }
+
+    Map<Double, String> heatMapColorHex = new HashMap<>();
+
+    if (iValues.size() > 1)
+    {
+        // calculate medianIndex of protein.getPeptideCharacteristics()
+        var count = iValues.size();
+        var medianIndex = 0;
+
+        if (count % 2 == 0)
+        {
+            medianIndex = (count / 2 - 1 + count / 2) / 2;
+        }
+        else
+        {
+            medianIndex = (count - 1) / 2;
+        }
+
+        // assign blue colors from median to last -> lighter to darker
+        Color one = Color.WHITE;
+        Color two = new Color(0, 81, 138);
+        List<Color> blueGradient = ColorGradient.createGradient(one, two, (count-medianIndex +1));
+
+        // assign blue colors from median to last -> lighter to darker
+        for (int i = medianIndex + 1; i < count; i++)
+        {
+
+            var peptideColor = blueGradient.get(i-medianIndex);
+            var peptideCharacteristic = peptideCharacteristics.get(i);
+            if (isIntensityView)
+            {
+                heatMapColorRGB.put(peptideCharacteristics.get(i).getIntensity(), peptideColor);
+                peptideCharacteristic.setIntensityColor("#" + Integer.toHexString(peptideColor.getRGB()).substring(2));

```

Comment:
Since we only set one color (either based on intensity or confidence), can do we need to have both intensityColor and confidenceColor on the objects? Could we just have a single color?

Reply:
yes, updated